### ISSUE TITLE
Add 'environment' concept to empire

### ISSUE THREAD

**Issue by phobologic** [remind101/empire#480]
This could be useful for all sorts of things, but right now the biggest benefit would be tagging & naming AWS resources with the appropriate environments.

**Comment by gileze33** [comment]
How do you guys normally manage this now? Do you run separate ECS clusters per environment?

**Comment by public** [comment]
I'm not entirely clear on what an environment even means here :) Is it just a way of grouping apps or enabling support for multiple clusters? Or maybe it's some kind of cluster wide global configuration thing?

**Comment by gileze33** [comment]
Say in our case we run stage and production copies of all of our apps & infrastructure - would you run both within the same cluster or have two separate empire installs?

**Comment by public** [comment]
I would probably want to run them within the same cluster unless there was some kind of security related reason to keep them separate.

**Comment by ejholmes** [comment]
We run separate Empire environments (separate ECS clusters) for production/staging/team specific staging for a couples reasons.

1. It is better for security to isolate bugs in staging from effecting production applications.
2. Because of the way DNS service discovery works within Empire, having separate environments makes it easier to reproduce a production environment in staging, since the apps will be the same name, and get the same CNAME record (e.g. "app.empire" is the location of "app" within any environemnt).

**Comment by rgabo** [comment]
tldr; we run **dev** and **prd** clusters/environments separately and anyone can also create their own temporary test environment in 10-15 minutes using Terraform.

Applications should not be aware of what *environment* they run in, because its a concern that should be external to them. Doing otherwise violates 12-factor application design, one of the core design goals of Empire. An *environment* is an artificial concept that should mean nothing else other than a set of configuration grouped together.

Money quote:

> In a twelve-factor app, env vars are granular controls, each fully orthogonal to other env vars. They are never grouped together as “environments”, but instead are independently managed for each deploy. This is a model that scales up smoothly as the app naturally expands into more deploys over its lifetime.

Check out http://12factor.net/config for more inspiration.

**Comment by phobologic** [comment]
Just to clarify - yeah, we don't suggest that apps themselves have any concept of the environment they run in.  If you have an 'if/then/else' switch in your code that changes how it behaves based on an environment, you're just asking for trouble.

Empire itself though is a different beast, since it's job is to manage environments.  The key is that it's unlikely we'd have Empire use the environment for much of anything other than naming/tagging resources, which would make it much easier to differentiate resources like services/tasks/task definitions when you run multiple empire environments in the same AWS account.

**Comment by rgabo** [comment]
That's a great point. We're already handling this with Terraform *environments*: we keep separate groups of resources that differ by a prefix: *dev*, *prd* and the like. Hashicorp Atlas actually calls this an *environment*. Each of our environments has its own ECS cluster, ASG, ELBs, SGs, etc.

The only tricky part is to know which API endpoint to talk to. Today I solve this with:

```
export EMPIRE_API_URL=$(terraform output empire.url)
```

which will just spit out the ELB URL for whatever environment I'm currently configured with Terraform (local or remote state, etc).

**Comment by phobologic** [comment]
@rgabo We handle that entirely in cloudformation/stacker here.  We assign an 'external domain' to each environment, and part of building that stack/environment is setting the appropriate DNS records in those domains, again via Cloudformation/stacker.

Can Terraform modify route53?  If so, that might be useful for you.  We know that our empire API endpoints are always empire.<whatever domain the environment is in>.

**Comment by rgabo** [comment]
This is what we're doing, only our naming convention is slightly different. At any time, I could be talking to dev-empire.our.tld or prd-empire.our.tld. Switching between those needs to happen outside of `emp` today.

`terraform output empire.url` is just a fancy way of asking Terraform: What is the FQDN of the Route53 record you created for the Empire ELB?

**Comment by phobologic** [comment]
Yeah - we've fallen back to using aliases.

`alias emp-prod="EMPIRE_API_URL=<prodURL> emp"`
`alias emp-stage="EMPIRE_API_URL=<stageURL> emp"`

**Comment by ejholmes** [comment]
Gonna go ahead and close this. The `--environment` flag was added in https://github.com/remind101/empire/commit/34e1b23748d5f50adf0202a78c5dd4106c17cc4d, and the new CloudFormation will use this as a prefix when creating CloudFormation stacks, and also sets it as an `environment` tag on the stack.
